File: contrib/clojure-package/test/org/apache/clojure_mxnet/infer/predictor_test.clj
;; Licensed to the Apache Software Foundation (ASF) under one or more
;; contributor license agreements.  See the NOTICE file distributed with
;; this work for additional information regarding copyright ownership.
;; The ASF licenses this file to You under the Apache License, Version 2.0
;; (the "License"); you may not use this file except in compliance with
;; the License.  You may obtain a copy of the License at
;;
;;    http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.
;;

(ns org.apache.clojure-mxnet.infer.predictor-test
  (:require [org.apache.clojure-mxnet.context :as context]
            [org.apache.clojure-mxnet.dtype :as dtype]
            [org.apache.clojure-mxnet.infer :as infer]
            [org.apache.clojure-mxnet.layout :as layout]
            [org.apache.clojure-mxnet.ndarray :as ndarray]
            [org.apache.clojure-mxnet.shape :as shape]
            [clojure.java.io :as io]
            [clojure.java.shell :refer [sh]]
            [clojure.string :refer [split]]
            [clojure.test :refer :all]
            [org.apache.clojure-mxnet.util :as util]))

(def model-dir "data/")
(def model-path-prefix (str model-dir "resnet-18/resnet-18"))
(def width 224)
(def height 224)

(when-not (.exists (io/file (str model-path-prefix "-symbol.json")))
  (sh "./scripts/infer/get_resnet_18_data.sh"))

(defn create-predictor []
  (let [descriptors [{:name "data"
                      :shape [1 3 height width]
                      :layout layout/NCHW
                      :dtype dtype/FLOAT32}]
        factory (infer/model-factory model-path-prefix descriptors)]
    (infer/create-predictor factory)))

(deftest predictor-test-with-ndarray
  (let [predictor (create-predictor)
        image-ndarray (-> "test/test-images/kitten.jpg"
                           infer/load-image-from-file
                           (infer/reshape-image width height)
                           (infer/buffered-image-to-pixels [3 width height])
                           (ndarray/expand-dims 0))
        predictions (infer/predict-with-ndarray predictor [image-ndarray])
        synset-file (-> (io/file model-path-prefix)
                        (.getParent)
                        (io/file "synset.txt"))
        synset-names (split (slurp synset-file) #"\n")
        [best-index] (ndarray/->int-vec (ndarray/argmax (first predictions) 1))
        best-prediction (synset-names best-index)]
    (is (= "n02123159 tiger cat" best-prediction))))

(deftest predictor-test
  (let [predictor (create-predictor)
        image-ndarray (-> "test/test-images/kitten.jpg"
                          infer/load-image-from-file
                          (infer/reshape-image width height)
                          (infer/buffered-image-to-pixels [3 width height])
                          (ndarray/expand-dims 0))
        predictions (infer/predict predictor [(ndarray/->vec image-ndarray)])
        synset-file (-> (io/file model-path-prefix)
                        (.getParent)
                        (io/file "synset.txt"))
        synset-names (split (slurp synset-file) #"\n")
        ndarray-preds (ndarray/array (first predictions) [1 1000])
        [best-index] (ndarray/->int-vec (ndarray/argmax ndarray-preds 1))
        best-prediction (synset-names best-index)]
    (is (= "n02123159 tiger cat" best-prediction))))

File: scala-package/infer/src/test/java/org/apache/mxnet/infer/javaapi/PredictorTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mxnet.infer.javaapi;

import org.apache.mxnet.javaapi.Context;
import org.apache.mxnet.javaapi.NDArray;
import org.apache.mxnet.javaapi.Shape;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.*;

public class PredictorTest {

    Predictor mockPredictor;

    @Before
    public void setUp() {
        mockPredictor = Mockito.mock(Predictor.class);
    }

    @Test
    public void testPredictWithFloatArray() {

        float tmp[][] = new float[1][224];
        for (int x = 0; x < 1; x++) {
            for (int y = 0; y < 224; y++)
                tmp[x][y] = (int) (Math.random() * 10);
        }

        float [][] expectedResult = new float[][] {{1f, 2f}};
        Mockito.when(mockPredictor.predict(tmp)).thenReturn(expectedResult);
        float[][] actualResult = mockPredictor.predict(tmp);

        Mockito.verify(mockPredictor, Mockito.times(1)).predict(tmp);
        Assert.assertArrayEquals(expectedResult, actualResult);
    }

    @Test
    public void testPredictWithNDArray() {

        float[] tmpArr = new float[224];
            for (int y = 0; y < 224; y++)
                tmpArr[y] = (int) (Math.random() * 10);

        NDArray arr = new org.apache.mxnet.javaapi.NDArray(tmpArr, new Shape(new int[] {1, 1, 1, 224}), new Context("cpu", 0));

        List<NDArray> inputList = new ArrayList<>();
        inputList.add(arr);

        NDArray expected = new NDArray(tmpArr, new Shape(new int[] {1, 1, 1, 224}), new Context("cpu", 0));
        List<NDArray> expectedResult = new ArrayList<>();
        expectedResult.add(expected);

        Mockito.when(mockPredictor.predictWithNDArray(inputList)).thenReturn(expectedResult);

        List<NDArray> actualOutput = mockPredictor.predictWithNDArray(inputList);

        Mockito.verify(mockPredictor, Mockito.times(1)).predictWithNDArray(inputList);

        Assert.assertEquals(expectedResult, actualOutput);
    }

    @Test
    public void testPredictWithIterablesNDArray() {

        float[] tmpArr = new float[224];
        for (int y = 0; y < 224; y++)
            tmpArr[y] = (int) (Math.random() * 10);

        NDArray arr = new org.apache.mxnet.javaapi.NDArray(tmpArr, new Shape(new int[] {1, 1, 1, 224}), new Context("cpu", 0));

        Set<NDArray> inputSet = new HashSet<>();
        inputSet.add(arr);

        NDArray expected = new NDArray(tmpArr, new Shape(new int[] {1, 1, 1, 224}), new Context("cpu", 0));
        List<NDArray> expectedResult = new ArrayList<>();
        expectedResult.add(expected);

        Mockito.when(mockPredictor.predictWithNDArray(inputSet)).thenReturn(expectedResult);

        List<NDArray> actualOutput = mockPredictor.predictWithNDArray(inputSet);

        Mockito.verify(mockPredictor, Mockito.times(1)).predictWithNDArray(inputSet);

        Assert.assertEquals(expectedResult, actualOutput);
    }

    @Test
    public void testPredictWithListOfFloatsAsInput() {
        List<List<Float>> input = new ArrayList<>();

        input.add(Arrays.asList(new Float[] {1f, 2f}));

        List<List<Float>> expectedOutput = new ArrayList<>(input);

        Mockito.when(mockPredictor.predict(input)).thenReturn(expectedOutput);

        List<List<Float>> actualOutput = mockPredictor.predict(input);

        Mockito.verify(mockPredictor, Mockito.times(1)).predict(input);

        Assert.assertEquals(expectedOutput, actualOutput);

    }
}